lib/icp/impl/vanilla_3d.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ */
+#include <cstddef>
+#include <iostream>
+#include <numeric>
+#include <vector>
+#include <cmath>
+#include <Eigen/Eigen>
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include <Eigen/SVD>
+#include <Eigen/Dense>
+
+#include "icp/impl/vanilla_3d.h"
+#include "algo/kdtree.h"
+#include "icp/geo.h"
+
+/* #name Vanilla */
+
+/* #desc The vanilla algorithm for ICP will match the point-cloud centers
+exactly and then iterate until an optimal rotation has been found. */
+
+namespace icp {
+    Vanilla_3d::Vanilla_3d([[maybe_unused]] const Config& config)
+        : ICP(), c(3, 0), current_cost_(std::numeric_limits<double>::max()) {}
+    Vanilla_3d::Vanilla_3d(): ICP(), c(3, 0), current_cost_(std::numeric_limits<double>::max()) {}
+    Vanilla_3d::~Vanilla_3d() {}
+
+    // Euclidean distance between two points
+    float Vanilla_3d::dist(const Eigen::Vector3d& pta, const Eigen::Vector3d& ptb) {
+        return (pta - ptb).norm();
+    }
+
+    NEIGHBOR Vanilla_3d::nearest_neighbor(const PointCloud& src, const PointCloud& dst) {
+        NEIGHBOR neigh;
+        neigh.distances.resize(src.cols());
+        neigh.indices.resize(src.cols());
+
+        // Build KDTree
+        std::vector<Eigen::Vector3d> dst_vec(dst.cols());
+        for (ptrdiff_t i = 0; i < dst.cols(); ++i) {
+            dst_vec[i] = dst.col(i);
+        }
+        icp::KdTree<Eigen::Vector3d> kdtree(dst_vec, 3);
```

Comment:
Are we rebuilding this every time? Doesn't that defeat the point? Maybe I'm not getting it.